### ISSUE TITLE
Observe different prediction result with same prediction dataset but different column sequence

### ISSUE THREAD

**Issue by wlssea** [dmlc/xgboost#636]
I am using R xgboost package. I ran prediction multiple times but got different result. Each time I was using the same prediction dataset with the same column name but the column sequence is different. For example, Age was in column1 in the 1st run but becomes column 3 in the 2nd run. Do I expect xgboost to give different prediction in this case? Do I need to keep prediction dataset exactly the same column sequence as in training dataset? 

Here is my code:

train_sparse <- sparse.model.matrix(Y~.-1, data=train)
test_sparse <- sparse.model.matrix(Y~.-1, data=test)
dtrain <- xgb.DMatrix(data = train_sparse, label = train$Y, missing=NA)
dtest <- xgb.DMatrix(data = test_sparse, label = test$Y, missing=NA)

watchlist <- list(test=dtest, train=dtrain)

param <- list(
  objective="binary:logistic",
  booster="gbtree",
  eta=0.7, 
  max.depth=6,
  subsample=0.7, 
  colsample_bytree=1
)
 
xgmodel <- xgb.train(params           = param,
                       data                = dtrain,
                       missing=NA,
                       nthread = 2,
                       nrounds             = 70,
                       verbose             = 2,
                       early.stop.round    = 30,
                       watchlist           = watchlist,
                       maximize            = TRUE,
                       eval.metric="auc"
)
  
#prediction
pred_sparse <- sparse.model.matrix(Y~.-1, data=preddata)
predresult<-predict(xgmodel,pred_sparse,missing=NA)

What I mean is if I change preddata column sequence, the predresult will change a lot. 
Do I have to keep preddata column sequence the same as train?

**Comment by tqchen** [comment]
We will need to make sure all index encoding remains the same, so you do need to make things consistent

**Comment by hannarud** [comment]
@tqchen - and what about additional columns in data? and what if some columns are absent?

**Comment by apoorva007** [comment]
@tqchen Is there a way I can get index encodings from the trained model?
My use case is that I want to serve interactive predictions from a trained xgb model. My trained model has categorical columns which I converted to one hot encoding using pandas.get_dummies.
The issue is that now my test data on which I want predictions also has to have same order of column encodings as the trained model dataframe. But since model will be trained before and predictions be taken after, this seems to be an additional overhead to maintain
